test/library.cpp

```diff
@@ -327,28 +411,292 @@ class LibraryTest : public ::testing::Test {
+
+    ASSERT_EQ(lib->migrateBookmarks(bookId1), 1);
+    allBookmarks = lib->getBookmarks(false);
+    ASSERT_EQ(allBookmarks.size(), 6);
+    EXPECT_EQ(allBookmarks[0].getBookId(), bookId1+"_date");
+    EXPECT_EQ(allBookmarks[1].getBookId(), "invalid-book-id");
     EXPECT_EQ(allBookmarks[2].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[3].getBookId(), bookId1+"_date");
+    EXPECT_EQ(allBookmarks[4].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[5].getBookId(), bookId1+"_date_flavour");
+
+    ASSERT_EQ(lib->migrateBookmarks(bookId1, bookId2), 0); // No more bookId1 bookmark
+
+    ASSERT_EQ(lib->migrateBookmarks(bookId1+"_date", bookId2), 2);
+    onlyValidBookmarks = lib->getBookmarks();
+    allBookmarks = lib->getBookmarks(false);
+
+    ASSERT_EQ(onlyValidBookmarks.size(), 5);
+    EXPECT_EQ(onlyValidBookmarks[0].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[1].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[2].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[3].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[4].getBookId(), bookId1+"_date_flavour");
+
+    ASSERT_EQ(allBookmarks.size(), 6);
+    EXPECT_EQ(allBookmarks[0].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[1].getBookId(), "invalid-book-id");
+    EXPECT_EQ(allBookmarks[2].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[3].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[4].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[5].getBookId(), bookId1+"_date_flavour");
+
+    ASSERT_EQ(lib->migrateBookmarks("invalid-book-id", bookId1), 1);
+
+    onlyValidBookmarks = lib->getBookmarks();
+    allBookmarks = lib->getBookmarks(false);
+
+    ASSERT_EQ(onlyValidBookmarks.size(), 6);
+    EXPECT_EQ(onlyValidBookmarks[0].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[1].getBookId(), bookId1);
+    EXPECT_EQ(onlyValidBookmarks[2].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[3].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[4].getBookId(), bookId2);
+    EXPECT_EQ(onlyValidBookmarks[5].getBookId(), bookId1+"_date_flavour");
+
+
+    ASSERT_EQ(allBookmarks.size(), 6);
+    EXPECT_EQ(allBookmarks[0].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[1].getBookId(), bookId1);
+    EXPECT_EQ(allBookmarks[2].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[3].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[4].getBookId(), bookId2);
+    EXPECT_EQ(allBookmarks[5].getBookId(), bookId1+"_date_flavour");
+}
+
+TEST_F(LibraryTest, MigrateBookmarkOlder)
+{
+    auto bookId = std::string("0c45160e-f917-760a-9159-dfe3c53cdcdd");
+
+    auto book = lib->getBookById(bookId);
+
+    auto validBookmark = createBookmark(book);
+    lib->addBookmark(validBookmark);
+
+    auto onlyValidBookmarks = lib->getBookmarks();
+
+    ASSERT_EQ(lib->getBestTargetBookId(validBookmark, kiwix::UPGRADE_ONLY), bookId+"_date");
+    ASSERT_EQ(lib->getBestTargetBookId(validBookmark, kiwix::ALLOW_DOWNGRADE), bookId+"_date");
+}
```

Comment:
1. `onlyValidBookmarks` is not used
2. `migrateBookmarks()` is not called/tested (inconsistency with the test name)